Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Vedad Haracic"
STUDENTID = "01353113"

---

# Assignment 2

## Step 0 (3 points)

Find two data sets online (from one or several sources) that would be interesting to combine:

* Each data set must have a different file format (either CSV, XML, or JSON).
* You may start from (but you are not limited to) the resource collections hinted at in Unit 2.
* Workable data-set sizes: The selected or extracted data sets should have thousands of entries (>= 1000), but not more than (<=) 10000 entries. If larger, use an excerpt from the original data set. Justify the extraction criteria using a Markup cell.
* Mind the additional restrictions on data sets (no curated ones, via Open Data portal, signal adoption of the data sets to the LEARN forum).


### Selected datasets

Data set 1: Leading Causes of Death in the US by States (format: CSV)

-Creator/ authors(s) of the data set: Centers for Disease Control and Prevention
-Names of the data repository and/or the Open Data portal used: http://www.data.gov/
-Dataset identifier (specific to the data repository):https://catalog.data.gov/dataset/age-adjusted-death-rates-for-the-top-10-leading-causes-of-death-united-states-2013
-Dataset publication year: December 16, 2015, Updated on: February 28, 2019
-Last accessed (datetime of accessing, obtaining a copy of the data set): 22.03.2019

Data set 2: Maryland Counties Socioeconomic Characteristics (format:JSON)

-Creator/ authors(s) of the data set: opendata.maryland.gov
-Names of the data repository and/or the Open Data portal used: http://www.data.gov/
-Dataset identifier (specific to the data repository): https://catalog.data.gov/dataset/maryland-counties-socioeconomic-characteristics-f7e16
-Dataset publication year: March 18, 2019
-Last accessed (datetime of accessing, obtaining a copy of the data set): 22.03.2019


**Project ideas**

Correlation between socioeconomic status and cause of death per state in the USA. Data for the project is incomplete for now, since I don't have the socioeconomic data for all states in the USA. If my team chooses to do my idea as a project, there is a source on poverty levels and other demographic statistics on: https://www.census.gov/ .


## Step 1  (4 points)

Detect the file format and file size (convert to KB or MB) of each data set, clearly documenting your actions (e.g. through commented code).

In [2]:
import os
#finding the data:
filePath1="data/data_notebook-1_leading_cause_of_death.csv"
filePath2="data/data_notebook-1_maryland.json"
#declaring file size variables:
fSize1 = os.path.getsize(filePath1)
fSize2 = os.path.getsize(filePath2)
#Displaying the results
print('File size of '+filePath1+' is: '+str(round((fSize1/1024),2)) + ' KBytes (rounded on 2 decimal points)') 
print('File size of '+filePath2+' is: '+str(round((fSize2/1024),2)) + ' KBytes (rounded on 2 decimal points)') 
#detecting the file format (we can see which extension the file has and see which format it is, but here is an automated version):
name1, ext1 = os.path.splitext(filePath1)
name2, ext2 = os.path.splitext(filePath2)
print("The format of the "+filePath1+" is: "+ ext1)
print("The format of the "+filePath2+" is: "+ ext2)



File size of data/data_notebook-1_leading_cause_of_death.csv is: 789.67 KBytes (rounded on 2 decimal points)
File size of data/data_notebook-1_maryland.json is: 95.56 KBytes (rounded on 2 decimal points)
The format of the data/data_notebook-1_leading_cause_of_death.csv is: .csv
The format of the data/data_notebook-1_maryland.json is: .json


### Summary of findings

**Data set 1**

Format: .csv, Size: 789.67 KB, table structure.

**Data set 2**

Format: JSON, Size: 95.56 KB, tree structure typicall for JSON


## Step 2  (4 points)

Validate the data sets according to the data format used: Are there any data-formatting issues? Hint: You may use online validators for CSV, XML, or JSON. Pls. clearly indicate any validators used and summarize their reports.

### Summary of findings

**Data set 1**

Validator used: CSVLint, https://csvlint.io
Validation results: https://csvlint.io/validation/5c98ce46aaa38a0004000056
This file is a valid csv file according to the validator.
You can use the validator in the last cell to validate CSV files using CSVLint API.

**Data set 2**

Validator used: https://jsonlint.com
Validation result: Valid JSON



## Step 3  (4 points)

Access the two data sources and inspect their content (e.g., using the Python recipes presented to you in the notebooks and tutorials). Describe the characteristics of the data sets (depending on the format).

In [3]:
print("Inspecting the CSV file: ")
print("---------------------------------------------------------------------")


#creating an empty list where rows will be stored
liste=[]
#opening the csv file:
with open ('./data/data_notebook-1_leading_cause_of_death.csv','r') as f:
    rows=f.readlines()
    print("Names of the collumns: \n "+rows[0])
    #creating a variable which counts the number of rows:
    counter=0
    print("Dispalying the first 10 rows: ")
    for row in rows:
        counter+=1
        liste.append(row.strip().split(','))
        if counter<10:
            print(row)
            
        
    print("This CSV file has "+str(counter)+" rows.")
variable_counter=0
string_counter=0
for i in liste:
    for j in i:
        variable_counter+=1
        if type(j)== str:
            string_counter+=1

print("Number of strings: ")            
print(string_counter)        
print("This csv has "+ str(variable_counter)+" variables.")

#Now I am going to inspect the JSON file:
print("\n")
print("\n")
print("Inspecting the JSON file:")
print("---------------------------------------------------------------------")

import json
with open ("./data/data_notebook-1_maryland.json","rb") as file:
    data=json.load(file)
    info=data['meta']['view']['description']
    column_names=[]
    for name in data['meta']['view']['columns']:
        #print(name['name'])
        column_names.append(name['name'])
print("Names of the columns:")
for i in column_names:
    print(i)
#Now I am going to explore the data inside:
json_data=[]
for j in data['data']:
    json_data.append(j)
json_dict={}
for i in json_data:
    
    json_dict=dict(zip(column_names,i))
print("---------------------------------------------------------------------")
Json_column_counter=0
#Displaying the 24th observation:
for k in json_dict:
    print("Column: "+str(k)+"--------> Value: "+str(json_dict[k]))
    Json_column_counter+=1
        
print("---------------------------------------------------------------------")
#Now I am going to count the variable type in the observation:
int_count=0
str_count=0
number_of_observations=0
none_counter=0
number_of_variables=0
for k in json_data:
    int_count=sum(isinstance(v, int) for v in k)
    str_count=sum(isinstance(v, str) for v in k)
    for v in k:
        number_of_variables+=1
        if v is None:
            none_counter+=1
    number_of_observations+=1
print("The number of type String in the observation:"+str(str_count))
print("The number of type int in the observation:"+str(int_count))
print("Number of None types across the observations:"+str(none_counter))
print("Number of variables across the observations:"+str(number_of_variables))
print("Number of columns per observation: "+str(Json_column_counter))
    

Inspecting the CSV file: 
---------------------------------------------------------------------
Names of the collumns: 
 Year,113 Cause Name,Cause Name,State,Deaths,Age-adjusted Death Rate

Dispalying the first 10 rows: 
Year,113 Cause Name,Cause Name,State,Deaths,Age-adjusted Death Rate

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,Alabama,2755,55.50

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,Alaska,439,63.10

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,Arizona,4010,54.20

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,Arkansas,1604,51.80

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,California,13213,32.00

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,Colorado,2880,51.20

2016,"Accidents (unintentional injuries) (V01-X59,Y85-Y86)",Unintentional injuries,Conne

### Summary of findings

**Data set 1**

Data set data_notebook-1_leading_cause_of_death.csv has 70206 variables across 10297 rows. These variables are all of type String.

**Data set 2**
Data set data_notebook-1_maryland.json
Print statements above show the collumn names, the number of string, int, and none types.

This JSON has multiple nestings:
meta{metadata is shown here, all column names and other information about this JSON is stored here}
data{24 observations in lists (just values)}
This JSON file has a small ammount of observations, but this is due to the number of counties in Maryland. 
In this file there are 1176 variables, many belonging to metadata. There are 24 observations in this file with 49 variables each.

This might seem small, but I could not find any datasets with data from all US states (there is a way to download them from the US Census Office, but they are in an excel format and have to be downloaded seperately for each year). When opened in text editor, this file has about 3500 lines. 

## Bonus work  (up to 4 points)

Automate the data-set validation (as described in Step 2) using a Python code fragment, for *one* data set. Use the single code cell below and comment your code.

In [4]:
# I am not sure if this work is being graded, but I will do it anyway, since I did not do this part in my homework from the previous semester
#Importing requests module for web requests
import requests
# Importing time module to delay the response (this is required according to API documentation)
import time
def CSVChecker(x):
    data = {
  'urls[]': x}
    user_agent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36'
    header={'User-Agent' : user_agent }
    #posting a file:
    request = requests.post('http://csvlint.io/package.json', data=data,headers=header)
    first=request.json()
    #getting the url for the verification results
    csvid=first['package']['url']+'.json'
    time.sleep(5)
    #getting the validation results and returning them:
    response=requests.get(csvid)
    result=response.json()
    return result
print(CSVChecker("https://catalog.data.gov/dataset/age-adjusted-death-rates-for-the-top-10-leading-causes-of-death-united-states-2013"))
#Somethimes CSVLint wont return the complete validation JSON file, if that happens, just run the cell again.
#CSVLint is still in testing

{'version': '0.1', 'licence': 'http://opendatacommons.org/licenses/odbl/', 'package': {'validations': []}}
